opentelemetry-user-events-trace/src/trace/exporter.rs

```diff
@@ -0,0 +1,167 @@
+use chrono::{DateTime, Utc};
+use eventheader::{FieldFormat, Level, Opcode};
+use eventheader_dynamic::{EventBuilder, EventSet, Provider};
+use opentelemetry::trace::SpanKind;
+use opentelemetry::trace::Status;
+use opentelemetry::{KeyValue, Value};
+use opentelemetry_sdk::error::{OTelSdkError, OTelSdkResult};
+use opentelemetry_sdk::trace::SpanData;
+use std::{
+    fmt::Debug,
+    sync::{Arc, Mutex},
+};
+
+/// UserEventsSpanExporter exports spans in EventHeader format to user_events tracepoint.
+pub(crate) struct UserEventsSpanExporter {
+    provider: Mutex<Provider>,
+    name: String,
+    event_set: Arc<EventSet>,
+}
+
+impl Debug for UserEventsSpanExporter {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "user_events span exporter (provider: {})", self.name)
+    }
+}
+
+use futures_util::future::BoxFuture;
+use opentelemetry_sdk::trace::SpanExporter;
+
+impl SpanExporter for UserEventsSpanExporter {
+    fn export(&mut self, batch: Vec<SpanData>) -> BoxFuture<'static, OTelSdkResult> {
+        // Using try_for_each is safe here as we do not expect the batch to have more than one span.
+        let result = batch.iter().try_for_each(|span| self.export_span(span));
+
+        Box::pin(async move { result })
+    }
+
+    fn shutdown(&mut self) -> OTelSdkResult {
+        // The explicit unregister() is done in shutdown()
+        // as it may not be possible to unregister during Drop
+        // as `Tracers` are typically *not* dropped.
+        if let Ok(mut provider) = self.provider.lock() {
+            provider.unregister();
+            Ok(())
+        } else {
+            Err(OTelSdkError::InternalFailure(
+                "Failed to acquire lock on provider".to_string(),
+            ))
+        }
+    }
+}
+
+impl UserEventsSpanExporter {
+    /// Create a new instance of the exporter
+    pub(crate) fn new(provider_name: &str) -> Result<Self, String> {
+        if provider_name.len() >= 234 {
+            return Err("Provider name must be less than 234 characters.".to_string());
+        }
+        if !provider_name
+            .chars()
+            .all(|c| c.is_ascii_alphanumeric() || c == '_')
+        {
+            return Err(
+                "Provider name must contain only ASCII letters, digits, and '_'.".to_string(),
+            );
+        }
+
+        let mut eventheader_provider = Provider::new(provider_name, &Provider::new_options());
+        let keyword = 1;
+        let event_set = eventheader_provider.register_set(Level::Informational, keyword);
+        let name = eventheader_provider.name().to_string();
```

Comment:
nit: Emit a debug log when this done with name of tracepoint

Reply:
Done.